R/pkg/R/DataFrame.R

```diff
@@ -1413,15 +1413,146 @@ setMethod("join",
+#' df1 <- jsonFile(sqlContext, path)
+#' df2 <- jsonFile(sqlContext, path2)
+#' merge(df1, df2) # Performs a Cartesian
+#' merge(df1, df2, by = "col1") # Performs an inner join based on expression
+#' merge(df1, df2, by.x = "col1", by.y = "col2", all.y = TRUE)
+#' merge(df1, df2, by.x = "col1", by.y = "col2", all.x = TRUE)
+#' merge(df1, df2, by.x = "col1", by.y = "col2", all.x = TRUE, all.y = TRUE)
+#' }
 setMethod("merge",
           signature(x = "DataFrame", y = "DataFrame"),
-          function(x, y, joinExpr = NULL, joinType = NULL, ...) {
-            join(x, y, joinExpr, joinType)
-          })
+          function(x, y, by.x = NULL, by.y = NULL, by = intersect(names(x), names(y)),
+                   all.x = FALSE, all.y = FALSE, all = FALSE,
+                   suffixes = c("_x","_y"), sort = FALSE, ... ) {
 
+            if (missing(x) | missing(y)) {
+              stop("x and y has to be specified")
+            }
+
+            if (length(suffixes) != 2) {
+              stop("suffixes must have length 2")
+            }
+
+            # join type is identified based on the values of all, all.x and all.y
+            # default join type is inner, according to R it should be natural but since it
+            # is not supported in spark inner join is used
+            joinType <- "inner"
+            if (all | (all.x & all.y)) {
+              joinType <- "outer"
+            } else if (all.x) {
+              joinType <- "left_outer"
+            } else if (all.y) {
+              joinType <- "right_outer"
+            }
+
+            # join expression is based on by.x, by.y if both by.x and by.y are not missing
+            # and by if by.x or by.y are missing or have different lengths
+            if (length(by.x) > 0 & length(by.x) == length(by.y)) {
+              joinX <- by.x
+              joinY <- by.y
+            } else if (length(by) > 0) {
+              joinX <- by
+              joinY <- by
+            } else {
+              stop("The intersection of dataframes is empty")
+            }
+
+            # sets alias for making colnames unique in dataframe 'x'
+            namesX <- names(x)
+            colsX <- lapply(seq_len(length(namesX)), function(i){
+              colNameX <- namesX[[i]]
+              colX <- getColumn(x, colNameX)
+              if (namesX[i] %in% by) {
+                newJoinX <- paste(colNameX, suffixes[1], sep = "")
+                colX <- alias(colX, newJoinX)
+              }
+              colX
+            })
+
+            # sets alias for making colnames unique in dataframe 'y'
+            namesY <- names(y)
+            colsY <- lapply(seq_len(length(namesY)), function(i){
+              colNameY <- namesY[[i]]
+              colY <- getColumn(y, colNameY)
+              if (namesY[i] %in% by) {
+                newJoinY <- paste(colNameY, suffixes[2], sep = "")
+                colY <- alias(colY, newJoinY)
+              }
+              colY
+            })
+
```

Comment:
The above two blocks of code can be extracted into a common private function and be reused.